File: auth-service-beta/src/main/java/com/halfi/auth/service/JwtService.java
package com.halfi.auth.service;

import com.halfi.auth.model.User;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.JwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;
import java.nio.charset.StandardCharsets;
import java.util.Date;

@Service
public class JwtService {

    // Читаем из application.yml: jwt.secret
    @Value("${jwt.secret}")
    private String secret;

    // Читаем из application.yml: jwt.access-expiry-ms (900000 = 15 мин)
    @Value("${jwt.access-expiry-ms}")
    private long accessExpiryMs;

    // Читаем из application.yml: jwt.refresh-expiry-ms (604800000 = 7 дней)
    @Value("${jwt.refresh-expiry-ms}")
    private long refreshExpiryMs;

    // ── Генерация ─────────────────────────────────────────────────────────────

    /**
     * Создаёт короткоживущий accessToken (15 мин).
     * Payload: sub=userId, email=user@mail.com, type=access
     */
    public String generateAccessToken(User user) {
        return buildToken(user, accessExpiryMs, "access");
    }

    /**
     * Создаёт долгоживущий refreshToken (7 дней).
     * Payload: sub=userId, type=refresh
     */
    public String generateRefreshToken(User user) {
        return buildToken(user, refreshExpiryMs, "refresh");
    }

    private String buildToken(User user, long expiryMs, String type) {
        Date now = new Date();
        Date expiration = new Date(now.getTime() + expiryMs);

        return Jwts.builder()
                .subject(user.getId().toString())          // sub = userId
                .claim("email", user.getEmail())           // доп. клейм
                .claim("type", type)                       // access | refresh
                .issuedAt(now)
                .expiration(expiration)
                .signWith(getSignKey())                    // HS256 подпись
                .compact();
    }

    // ── Валидация ─────────────────────────────────────────────────────────────

    /**
     * Парсит и валидирует токен.
     * Бросает JwtException если токен невалиден или просрочен.
     */
    public Claims validateToken(String token) {
        // parseSignedClaims — в jjwt 0.12.x (раньше было parseClaimsJws)
        return Jwts.parser()
                .verifyWith(getSignKey())
                .build()
                .parseSignedClaims(token)
                .getPayload();
    }

    /**
     * Возвращает userId из токена без проброса исключения.
     * Возвращает null если токен невалиден.
     */
    public String extractUserId(String token) {
        try {
            return validateToken(token).getSubject();
        } catch (JwtException e) {
            return null;
        }
    }

    /**
     * Возвращает email из токена.
     */
    public String extractEmail(String token) {
        try {
            return validateToken(token).get("email", String.class);
        } catch (JwtException e) {
            return null;
        }
    }

    /**
     * Проверяет что токен валиден (не бросает исключение).
     */
    public boolean isTokenValid(String token) {
        try {
            validateToken(token);
            return true;
        } catch (JwtException e) {
            return false;
        }
    }

    // ── Приватный хелпер ──────────────────────────────────────────────────────

    private SecretKey getSignKey() {
        // Keys.hmacShaKeyFor требует минимум 32 байта для HS256
        byte[] keyBytes = secret.getBytes(StandardCharsets.UTF_8);
        return Keys.hmacShaKeyFor(keyBytes);
    }
}

File: auth-service/src/main/kotlin/com/halfi/service/JwtService.kt
package com.halfi.auth.service

import com.halfi.auth.entity.User
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.SignatureAlgorithm
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Service
import java.util.*

@Service
class JwtService {
    // В будущем вынесем в application.yml (и он должен быть минимум 256 бит)
    private val secret = "your-very-secure-secret-key-that-must-be-very-long" 
    private val key = Keys.hmacShaKeyFor(secret.toByteArray())

    fun generateAccessToken(user: User): String {
        return Jwts.builder()
            .setSubject(user.email)
            .claim("userId", user.id.toString())
            .claim("role", user.role)
            .setIssuedAt(Date())
            .setExpiration(Date(System.currentTimeMillis() + 1000 * 60 * 15)) // 15 минут
            .signWith(key, SignatureAlgorithm.HS256)
            .compact()
    }

    fun generateRefreshToken(user: User): String {
        return Jwts.builder()
            .setSubject(user.email)
            .claim("userId", user.id.toString())
            .setIssuedAt(Date())
            .setExpiration(Date(System.currentTimeMillis() + 1000L * 60 * 60 * 24 * 30)) // 30 дней
            .signWith(key, SignatureAlgorithm.HS256)
            .compact()
    }
}